cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/SemanticSSA.qll

```diff
@@ -70,6 +70,21 @@ predicate semBackEdge(SemSsaPhiNode phi, SemSsaVariable inp, SemSsaReadPositionP
   // Conservatively assume that every edge is a back edge if we don't have dominance information.
   (
     phi.getBasicBlock().bbDominates(edge.getOrigBlock()) or
+    trimmedReachable(phi.getBasicBlock(), edge.getOrigBlock()) or
```

Comment:
Replace `trimmedReachable` with `trimmedEdges+`. Also, this condition appears necessary, but not sufficient, to identify irreducible SCC edges. This is going to mark too many edges as back edges: consider e.g. the phi node in
```
if (..) x = 1; else x = 2;
```
Without a surrounding loop this phi node should not have any incoming back-edges.
Use the following instead
```suggestion
    irreducibleSccEdge(phi.getBasicBlock(), edge.getOrigBlock()) or
```
and
```
private predicate irreducibleSccEdge(SemBasicBlock b1, SemBasicBlock b2) {
  trimmedEdges(b1, b2) and trimmedEdges+(b2, b1)
}
```
Also, style nit pick: replace `trimmedEdges` with `trimmedEdge` - we generally prefer singular to describing a predicate member tuple rather than the entire set.

Reply:
And then some qldoc explaining things about reducible vs. irreducible CFGs is probably in order.